src/libcore/iter.rs

```diff
@@ -115,7 +119,10 @@ pub trait Iterator {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     fn count(self) -> usize where Self: Sized {
-        self.fold(0, |cnt, _x| cnt + 1)
+        self.fold(0, |cnt, _| match cnt.checked_add(1) {
+            Some(c) => c,
+            None => panic!("overflow while counting the elements of an iterator"),
```

Comment:
I don't like this notion TBH. In C this is what you'd expect: The function will go horribly wrong if passed the wrong parameters. But in Rust making errors silent is not-so-much the style. If someone needs the more performant variant of this, maybe they should code their own version? One alternative I can think of would be to provide an extra `count_unchecked` function.